### PR TITLE
add dns nameserver check to prevent only localhost dns nameserver

### PR DESCRIPTION
**Pull request by aiwantaozi**
Related Issue: https://github.com/rancher/rancher/issues/9562
Description: Add dns checker to prevent only localhost dns namserver in the resolv.conf in the metadata, because it use bridge network and config dns search when create, so docker will not automatically add 8.8.8.8 to the resolv.conf, more details in the issue.

@yasker  could you help review this PR.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -171,6 +181,28 @@ func appMain(ctx *cli.Context) error {
...
+		if !strings.HasPrefix(text, "nameserver") {
+			continue
+		}
+
+		if !strings.Contains(text, "127.0.0.1") && !strings.Contains(text, "127.0.1.1") {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("no available nameserver, please check file /etc/resolv.conf on the host")
```
Comment (yasker) [inline main.go]: I think you meant `valid` nameserver here.

**Comment by yasker** [comment]
Do we need to do the same thing with DNS service?

**Comment by aiwantaozi** [comment]
@yasker The dns depend on metadata, if the metadata could get up, then the dns is OK, I think we don't need this in the dns.

**Comment by yasker** [comment]
The whole `127.0.0.1/8` block is for loopback. We should exclude it.

**Comment by aiwantaozi** [comment]
@yasker Just update to prevent all the loopback ips, could you help review, thanks

File: main.go [hunk 2]
```diff
@@ -34,6 +37,11 @@ const (
 	DEFAULT_KEY = "default"
 )
 
+const (
+	resolvConfLocation = "/etc/resolv.conf"
+	ipLocalhost        = `^nameserver\s+((127\.([0-9]{1,3}\.){2}[0-9]{1,3})|(::1)$)`
```
Comment (yasker) [inline main.go]: You can do `^nameserver\s+((127(\.[0-9]{1,3}){3})|(::1)$)` instead.

File: main.go [hunk 3]
```diff
@@ -10,19 +11,21 @@ import (
...
 	"os/signal"
+	"regexp"
 	"sort"
 	"strconv"
 	"strings"
 	"sync"
 	"syscall"
 	"time"
 
+	_ "net/http/pprof"
```
Comment (yasker) [inline main.go:22]: Why?

**Review by yasker — CHANGES_REQUESTED** [review]

File: main.go (hunk 2 above)
Comment (aiwantaozi) [reply main.go]: @yasker Hi Yasker, 127.0.0.1, the last number not has the dot.

Comment (yasker) [reply main.go]: You read it wrong.

You grouped it as ,`127.[XXX.]{repeat twice}XXX`

What you should do it as `127[.XXX]{repeat three times}`, as in my regex.

Comment (aiwantaozi) [reply main.go]: Just updated to use ^nameserver\s+((127(\.[0-9]{1,3}){3})|(::1)$), thanks

File: main.go (hunk 3 above)
Comment (yasker) [reply main.go:22]: This one missed?

File: main.go [hunk 4]
```diff
@@ -10,19 +11,21 @@ import (
...
+
 	"github.com/Sirupsen/logrus"
 	"github.com/codegangsta/cli"
 	"github.com/golang/gddo/httputil"
 	"github.com/gorilla/mux"
 	yaml "gopkg.in/yaml.v2"
-	_ "net/http/pprof"
```
Comment (aiwantaozi) [inline main.go:25]: @yasker  This line used to place here, and the vscode auto format it, separated the init file import package and the real use package.

**Review by yasker — APPROVED** [review]

**Comment by alena1108** [comment]
Closing as it was fixed on the rancher/agent side.

**Pull request closed without merging**